lib/streamlit/widgets.py

```diff
@@ -73,15 +75,60 @@ def get_widget_value(self, widget_id: str) -> Optional[Any]:
 
     def set_state(self, widget_states: WidgetStates) -> None:
         """Copy the state from a WidgetStates protobuf into our state dict."""
-        self._state = {}
+        self._widget_states = {}
         for wstate in widget_states.widgets:
-            self._state[wstate.id] = wstate
+            self._widget_states[wstate.id] = wstate
+
+    def add_callback(
+        self,
+        widget_id: str,
+        deserializer: Callable[..., Any],
+        callback: Callable[..., None],
+    ) -> None:
+        """Add a callback that will be called immediately before the app's
+        next rerun if the given widget's value has changed.
+        """
+        self._widget_callbacks[widget_id] = callback
+        self._widget_deserializers[widget_id] = deserializer
+
+    def get_callback(self, widget_id: str) -> Optional[Callable[..., None]]:
+        return self._widget_callbacks.get(widget_id, None)
+
+    def call_callbacks(self, new_widget_states: WidgetStates) -> None:
+        """Call all callbacks based on changes in the widget state. We are
+        assuming that widgets send the same state value each time if unchanged.
+        """
+        for new_state in new_widget_states.widgets:
+            callback = self._widget_callbacks.get(new_state.id, None)
+            deserializer = self._widget_deserializers.get(new_state.id, None)
+            if callback is None:
+                # The widget does not have an on_change callback.
+                continue
```

Comment:
micronit: move the `deserializer` dict lookup till after the `if callback is None: continue` clause so that we don't do unnecessary work before bailing.